src/Microsoft.Data/DataFrameTable.cs

```diff
@@ -0,0 +1,133 @@
+            for (int ii = 0; ii < ColumnCount; ii++)
+            {
+                ret.Add(Column(ii)[rowIndex]);
+            }
+            return ret;
+        }
+
+        public void InsertColumn<T>(int columnIndex, IEnumerable<T> column, string columnName)
+            where T : struct
+        {
+            column = column ?? throw new ArgumentNullException(nameof(column));
+            if (columnIndex < 0 || columnIndex > _columns.Count)
+            {
+                throw new ArgumentException($"Invalid columnIndex {columnIndex} passed into Table.AddColumn");
+            }
+            BaseDataFrameColumn newColumn = new PrimitiveDataFrameColumn<T>(columnName, column);
+            InsertColumn(columnIndex, newColumn);
+        }
+        public void InsertColumn(int columnIndex, BaseDataFrameColumn column)
+        {
+            column = column ?? throw new ArgumentNullException(nameof(column));
+            if (columnIndex < 0 || columnIndex > _columns.Count)
+            {
+                throw new ArgumentException($"Invalid columnIndex {columnIndex} passed into Table.AddColumn");
+            }
+            if (RowCount > 0 && column.Length != RowCount)
+            {
+                throw new ArgumentException($"Column's length {column.Length} must match Table's length {RowCount}");
+            }
+            if (_columnNameToIndexDictionary.ContainsKey(column.Name))
+            {
+                throw new ArgumentException($"Table already contains a column called {column.Name}");
+            }
+            RowCount = column.Length;
+            _columnNames.Insert(columnIndex, column.Name);
+            _columnNameToIndexDictionary[column.Name] = columnIndex;
+            _columns.Insert(columnIndex, column);
+            ColumnCount++;
+        }
+        public void SetColumn(int columnIndex, BaseDataFrameColumn column)
+        {
+            column = column ?? throw new ArgumentNullException(nameof(column));
+            if (columnIndex < 0 || columnIndex >= ColumnCount)
+            {
+                throw new ArgumentException($"Invalid columnIndex {columnIndex} passed in to Table.SetColumn");
+            }
+            if (RowCount > 0 && column.Length != RowCount)
+            {
+                throw new ArgumentException($"Column's length {column.Length} must match table's length {RowCount}");
+            }
+            if (_columnNameToIndexDictionary.ContainsKey(column.Name))
+            {
+                throw new ArgumentException($"Table already contains a column called {column.Name}");
+            }
+            _columnNameToIndexDictionary.Remove(_columnNames[columnIndex]);
+            _columnNames[columnIndex] = column.Name;
+            _columnNameToIndexDictionary[column.Name] = columnIndex;
+            _columns[columnIndex] = column;
+        }
+        public void RemoveColumn(int columnIndex)
+        {
+            _columnNameToIndexDictionary.Remove(_columnNames[columnIndex]);
```

Comment:
Should you validate for -1

Reply:
Nope I think. _columnName[columnIndex] will throw the exception I want

Reply:
It's a bad practice not to check parameters and delegate their validation.